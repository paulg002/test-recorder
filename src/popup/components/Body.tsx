import * as React from 'react';
import type { Block } from '../../types';
import { RecState } from '../../constants';
import CodeDisplay from './CodeDisplay';
import LandingBox from './LandingBox';

export interface BodyProps {
  isValidTab: boolean,
  recStatus: RecState,
  codeBlocks: Block[],
  destroyBlock: (index: number) => void,
  moveBlock: (dragIdx: number, dropIdx: number) => void,
}

export default ({
  recStatus,
  codeBlocks,
  isValidTab,
  destroyBlock,
  moveBlock,
}: BodyProps) => (
  <div id="body">
    {recStatus === RecState.OFF && <LandingBox isValidTab={isValidTab} />}
    {recStatus !== RecState.OFF && <CodeDisplay codeBlocks={codeBlocks} destroyBlock={destroyBlock} moveBlock={moveBlock} />}
  </div>
);
