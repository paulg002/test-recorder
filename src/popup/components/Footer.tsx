import * as React from 'react';
import ToggleButton from './ToggleButton';
import ResetButton from './ResetButton';
import ClipboardButton from './ClipboardButton';
import { ControlAction, RecState } from '../../constants';

export interface FooterProps {
  isValidTab: boolean,
  recStatus: RecState,
  handleToggle: (action: ControlAction) => void,
  copyToClipboard: () => Promise<void>,
}

export default ({
  isValidTab,
  recStatus,
  handleToggle,
  copyToClipboard,
} : FooterProps) => (
  <div id="footer">
    <ToggleButton recStatus={recStatus} handleToggle={handleToggle} isValidTab={isValidTab} />
    {recStatus === RecState.PAUSED && <ResetButton handleToggle={handleToggle} />}
    {recStatus === RecState.PAUSED && <ClipboardButton copyToClipboard={copyToClipboard} />}
  </div>
);
