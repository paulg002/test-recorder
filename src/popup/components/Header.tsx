import * as React from 'react';
import InfoButton from './InfoButton';

export interface HeaderProps {
  toggleInfoDisplay: () => void,
  shouldInfoDisplay: boolean,
}

export default ({ shouldInfoDisplay, toggleInfoDisplay }: HeaderProps) => (
  <div id="header">
    <h1 id="title">Test recorder</h1>
    <InfoButton shouldInfoDisplay={shouldInfoDisplay} toggleInfoDisplay={toggleInfoDisplay} />
  </div>
);
