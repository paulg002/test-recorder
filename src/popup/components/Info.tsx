import * as React from 'react';

export default () => {
  const handleClick = (): void => {
    window.open('https://github.com/paulg002/test-recorder');
  };

  return (
    <div id="infobox">
      <h1>Hello.</h1>
      <br />
      <p>Thanks for checking out this developer tool.</p>
      <br />
      <p>
        We want to make automated testing very simple. Just load your app, click record on the main
        page and use your app as any user would. Once you&apos;re done, stop recording and copy the
        generated test code to your clipboard to use in your Cypress tests.
      </p>
      <br />
      <p>
        This tool was developed by the folks at paulg002 after realizing the need and community
        interest in the tool. It works by analyzing user activity in your app (clicks, keypresses,
        routes, and more) and then turning those interactions into Cypress code used for testing
        purposes.
      </p>
      <br />
      <p>
        Here is a link to the project&apos;s&nbsp;
        <span role="link" tabIndex={0} onClick={handleClick} onKeyDown={handleClick}>
          <a href="https://github.com/paulg002/test-recorder">Github Repo</a>
        </span>
        . Give it a star!!
      </p>
    </div>
  );
};
