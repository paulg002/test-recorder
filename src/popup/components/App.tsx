import * as React from 'react';
import Header from './Header';
import Info from './Info';
import Footer from './Footer';
import Body from './Body';
import type { Block, ActionWithPayload } from '../../types';
import { ControlAction, RecState } from '../../constants';
import '../../assets/styles/styles.scss';

export default () => {
  const [recStatus, setRecStatus] = React.useState<RecState>(RecState.OFF);
  const [codeBlocks, setCodeBlocks] = React.useState<Block[]>([]);
  const [shouldInfoDisplay, setShouldInfoDisplay] = React.useState<boolean>(false);
  const [isValidTab, setIsValidTab] = React.useState<boolean>(true);

  const startRecording = (): void => {
    setRecStatus(RecState.ON);
  };
  const stopRecording = (): void => {
    setRecStatus(RecState.PAUSED);
  };
  const resetRecording = (): void => {
    setRecStatus(RecState.OFF);
    setCodeBlocks([]);
  };

  React.useEffect((): void => {
    chrome.storage.local.get(['status', 'codeBlocks'], result => {
      if (result.codeBlocks) setCodeBlocks(result.codeBlocks);
      if (result.status === RecState.ON) setRecStatus(RecState.ON);
      else if (result.status === RecState.PAUSED) setRecStatus(RecState.PAUSED);
    });
    chrome.tabs.query({ active: true, currentWindow: true }, activeTab => {
      if (activeTab[0].url.startsWith('chrome://')) setIsValidTab(false);
    });
  }, []);

  React.useEffect((): () => void => {
    function handleMessageFromBackground({ type, payload }: ActionWithPayload): void {
      setShouldInfoDisplay(false);
      if (type === ControlAction.START && isValidTab) startRecording();
      else if (type === ControlAction.STOP) stopRecording();
      else if (type === ControlAction.RESET) resetRecording();
      else if (type === ControlAction.PUSH) setCodeBlocks(blocks => [...blocks, payload]);
    }
    chrome.runtime.onMessage.addListener(handleMessageFromBackground);
    return () => {
      chrome.runtime.onMessage.removeListener(handleMessageFromBackground);
    };
  }, []);

  const handleToggle = (action: ControlAction): void => {
    if (shouldInfoDisplay) setShouldInfoDisplay(false);
    if (action === ControlAction.START) startRecording();
    else if (action === ControlAction.STOP) stopRecording();
    else if (action === ControlAction.RESET) resetRecording();
    chrome.runtime.sendMessage({ type: action });
  };

  const toggleInfoDisplay = (): void => {
    setShouldInfoDisplay(should => !should);
  };

  const copyToClipboard = async (): Promise<void> => {
    try {
      let toBeCopied: string = '';
      for (let i = 0; i !== codeBlocks.length; i += 1) {
        toBeCopied += codeBlocks[i].value.concat('\n');
      }
      await navigator.clipboard.writeText(toBeCopied);
    } catch (error) {
      throw new Error(error);
    }
  };

  const destroyBlock = (index: number): void => {
    setCodeBlocks(prevBlocks => prevBlocks.filter((block, i) => i !== index));
    chrome.runtime.sendMessage({
      type: ControlAction.DELETE,
      payload: index,
    });
  };

  const moveBlock = (dragIdx: number, dropIdx: number): void => {
    const temp = [...codeBlocks];
    const dragged = temp.splice(dragIdx, 1)[0];
    temp.splice(dropIdx, 0, dragged);
    setCodeBlocks(temp);
    chrome.runtime.sendMessage({
      type: ControlAction.MOVE,
      payload: { dragIdx, dropIdx },
    });
  };

  return (
    <div id="App">
      <Header shouldInfoDisplay={shouldInfoDisplay} toggleInfoDisplay={toggleInfoDisplay} />
      {
        (shouldInfoDisplay
          ? <Info />
          : (
            <Body
              codeBlocks={codeBlocks}
              recStatus={recStatus}
              isValidTab={isValidTab}
              destroyBlock={destroyBlock}
              moveBlock={moveBlock}
            />
          )
        )
      }
      <Footer
        isValidTab={isValidTab}
        recStatus={recStatus}
        handleToggle={handleToggle}
        copyToClipboard={copyToClipboard}
      />
    </div>
  );
};
