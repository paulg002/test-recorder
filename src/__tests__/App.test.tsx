import * as React from 'react';
import { shallow } from 'enzyme';
import App from '../popup/components/App';
import Header from '../popup/components/Header';
import Body from '../popup/components/Body';
import Footer from '../popup/components/Footer';
import Info from '../popup/components/Info';
import '../setupTests';

describe('App', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<App />);
  });
  it('Should match snapshot, render Header, Body, and Footer', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.exists(Header)).toBe(true);
    expect(wrapper.exists(Body)).toBe(true);
    expect(wrapper.exists(Footer)).toBe(true);
    expect(wrapper.exists(Info)).toBe(false);
  });
});
