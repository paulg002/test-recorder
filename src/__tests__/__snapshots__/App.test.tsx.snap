// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should match snapshot, render Header, Body, and Footer 1`] = `
<div
  id="App"
>
  <Component
    shouldInfoDisplay={false}
    toggleInfoDisplay={[Function]}
  />
  <Component
    codeBlocks={Array []}
    destroyBlock={[Function]}
    isValidTab={true}
    moveBlock={[Function]}
    recStatus={1}
  />
  <Component
    copyToClipboard={[Function]}
    handleToggle={[Function]}
    isValidTab={true}
    recStatus={1}
  />
</div>
`;
